timelock-impl/src/main/java/com/palantir/atlasdb/timelock/watch/LockWatch.java

```diff
@@ -0,0 +1,57 @@
+/*
+ * (c) Copyright 2019 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.timelock.watch;
+
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicLong;
+
+import com.google.common.collect.Maps;
+import com.palantir.common.streams.KeyedStream;
+import com.palantir.lock.LockDescriptor;
+
+class LockWatch {
+    private Map<LockDescriptor, LockIndexState> indexStates;
+    private Map<LockDescriptor, AtomicLong> counters;
+
+    LockWatch(Set<LockDescriptor> lockDescriptors) {
+        this.indexStates = KeyedStream.of(lockDescriptors)
+                .map(LockIndexState::createDefaultForLockDescriptor)
+                .collectTo(Maps::newConcurrentMap);
+        this.counters = KeyedStream.of(lockDescriptors)
+                .map(unused -> new AtomicLong())
+                .collectTo(Maps::newConcurrentMap);
+    }
+
+    void registerLock(LockDescriptor descriptor) {
+        indexStates.computeIfPresent(
+                descriptor,
+                (unused, oldState) -> oldState.withLockSequence(counters.get(descriptor).incrementAndGet()));
+    }
+
+    void registerUnlock(LockDescriptor descriptor) {
+        indexStates.computeIfPresent(
+                descriptor,
+                (unused, oldState) -> oldState.withUnlockSequence(counters.get(descriptor).incrementAndGet()));
+    }
+
+    LockWatchState getState() {
+        return ImmutableLockWatchState.builder()
```

Comment:
The guarantees we get are enough, I think:

> The view's returned iterator is a "weakly consistent" iterator that will never throw ConcurrentModificationException, and guarantees to traverse elements as they existed upon construction of the iterator, and may (but is not guaranteed to) reflect any modifications subsequent to construction.

It's ok to read old values here - the dangerous case is if you miss a lock, since an unlock makes behaviour more permissive on clients. Users acquire locks before they can get their commit timestamp. However, by this point the calling transaction will have gotten their start, so it is not wrong to read the cached value - since any associated updates would not be visible to the calling transaction.